Feat/ccip 2466 HomeChain config syncer

asoliman92 opened the pull request:
## Motivation

Plugins need to refresh HomeChain config regularly to know which oracles support which chains and the latest F for each chain.

## Solution

* Add HomeChainConfigPoller which polls chain configs and maps them to internal configs
* Commit plugin uses HomeChainConfigPoller to access required data

asoliman92 committed:
WIP Add HomeChainConfig to CommitPlugin

asoliman92 committed:
Integrate HomeChainConfig into commit plugin

* Removed unused fields from CommitPluginConfig
* Added CommitConsensusConfig specifically for configs that we'll need to make consensus on before reaching an outcome
* Added `HomeChainConfig` field to `CommitPluginObservation`

Important Note:
Copied all cciptypes over from chainlink-common to be able to test freely. Once the draft is approved I'll move them back and depend on chainlink-common directly.

asoliman92 committed:
Mapping OnChain config to HomeChainConfig

* Implemented conversion between OnChainConfig and HomeChainConfig
* Added mock reader
* Added tests to test conversion and polling

asoliman92 committed:
Integrate HomeChainConfigPoller into commit plugin

* Use HomeChainPoller interface to get the homechain config when needed during the plugin run
* Removed PluginConfig from `CommitPluginObservation` and added `ConsensusObservation` that includes all the fields that we need to run consensus on during the outcome phase.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/common_types.go:
```diff
@@ -0,0 +1,84 @@
+package cciptypes
```
Added P2PID type

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/common_types_test.go:
```diff
@@ -0,0 +1,41 @@
+package cciptypes
```
Copied from Chainlink-common, ignore

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/config.go:
```diff
@@ -0,0 +1,173 @@
+package cciptypes
```
Important changes next..

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/config.go:
```diff
@@ -0,0 +1,173 @@
+	// DestChain is the ccip destination chain configured for the commit plugin DON.
+	DestChain ChainSelector `json:"destChain"`
+
+	// TODO: delete
```
Removing these..

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/config.go:
```diff
@@ -0,0 +1,173 @@
+	//FChain map[ChainSelector]int `json:"fChain"`
+
+	// PricedTokens is a list of tokens that we want to submit price updates for.
+	PricedTokens []types.Account `json:"pricedTokens"`
```
I think this should be part of `ObservationConsensus` struct not the plugin config. It seems that it will be changing and we need to reach consensus on it between observers in DON.
CC @dimkouv @makramkd

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/config.go:
```diff
@@ -0,0 +1,173 @@
+	Reads []ChainSelector `json:"reads"`
+}
+
+// TODO: Add all following
```
All the following till the end of file is new.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/config.go:
```diff
@@ -0,0 +1,173 @@
+	FChain  uint8     `json:"fChain"`
+	Config  []byte    `json:"config"`
+}
+type OnChainCapabilityConfig struct {
```
@makramkd Can you double check that I'm doing this correctly?

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/generic_types.go:
```diff
@@ -0,0 +1,101 @@
+package cciptypes
```
Copied from Chainlink-common, ignore.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/generic_types_test.go:
```diff
@@ -0,0 +1,145 @@
+package cciptypes
```
Copied from Chainlink-common, ignore.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/interfaces.go:
```diff
@@ -0,0 +1,62 @@
+package cciptypes
```
Adding `HomeChainPoller` interface

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/plugin_commit_types.go:
```diff
@@ -0,0 +1,146 @@
+package cciptypes
```
Important changes here.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/plugin_commit_types.go:
```diff
@@ -0,0 +1,146 @@
+	GasPrices            []GasPriceChain      `json:"gasPrices"`
+	TokenPrices          []TokenPrice         `json:"tokenPrices"`
+	MaxSeqNums           []SeqNumChain        `json:"maxSeqNums"`
+	ConsensusObservation ConsensusObservation `json:"consensusObservation"`
```
Removed PluginConfig and using `ConsensusObservation` instead that has only fields relevant for observations that we need to reach consensus on.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/plugin_commit_types_test.go:
```diff
@@ -0,0 +1,96 @@
+package cciptypes
```
Copied from Chainlink-common, ignore.

asoliman92 commented on core/services/ocr3/plugins/ccip/ccipocr3-dont-merge/plugin_execute_types.go:
```diff
@@ -0,0 +1,112 @@
+package cciptypes
```
Copied from Chainlink-common, ignore.

asoliman92 commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -27,28 +28,24 @@ type Plugin struct {
-		knownSourceChains = knownSourceChains.Union(mapset.NewSet(inf.Reads...))
-	}
+	// Start polling the home chain config in the background every 6 minutes
+	go homeChainPoller.StartPolling(ctx, 360*time.Second)
```
I think generally we shouldn't be able to create the plugin without having `homeChainConfig` ready, it's a field inside `homeChainPoller`. These checks should happen in the factory before even starting the plugin IMO. Or if we're moving the Poller to be on JobSpec level that will be something else as well.

asoliman92 commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -163,7 +166,10 @@ func (p *Plugin) ValidateObservation(_ ocr3types.OutcomeContext, _ types.Query,
+	homeChainConfig := p.homeChainPoller.GetConfig()
+
+	// TODO: This doesn't compare consensus observation with the home chain config's NodeSupportedChains
+	if err := validateObserverReadingEligibility(ao.Observer, obs.NewMsgs, homeChainConfig.NodeSupportedChains); err != nil {
```
CC @dimkouv

asoliman92 committed:
Fix tests

dimkouv commented on core/services/ocr3/plugins/ccip/internal/mocks/home_chain_poller.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+type HomeChainPollerMock struct {
+	*mock.Mock
```
`Mock` is not used.
Is the intention to use `Mock` or the statically defined config and `Close` to always return `nil`? IMO we should use `Mock`

asoliman92 replied:
Agreed 👍

dimkouv commented on core/services/ocr3/plugins/ccip/internal/mocks/home_chain_reader.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+type HomeChainContractReaderMock struct {
+	*mock.Mock
```
Same here (comment on `Mock`)

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+	ticker := time.NewTicker(interval)
+	defer ticker.Stop()
+
+	log.Println("Start Polling HomeChainConfig")
```
We should use `logger` similar to plugins.

dimkouv replied:
And use proper log levels e.g. info or debug here.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+			if err != nil {
+				log.Println("Error polling DB:", err)
+			}
+			r.homeChainConfig = homeConfig
```
There is a race condition on this variable.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+	return r.homeChainConfig
+}
+
+func (r *HomeChainConfigPoller) GetSupportedChains(oracleID commontypes.OracleID) mapset.Set[cciptypes.ChainSelector] {
```
This method is not used.

asoliman92 replied:
It's used in `Observation` step

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+		case <-ticker.C:
+			homeConfig, err := r.fetchLatestConfig(ctx)
+			if err != nil {
+				log.Println("Error polling DB:", err)
```
If you get an error you overwrite with an empty config.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+	var onChainCapabilityConfig []cciptypes.OnChainCapabilityConfig
+	err := r.homeChainReader.GetLatestValue(ctx, "CCIPCapabilityConfiguration", "getAllChainConfigs", nil, &onChainCapabilityConfig)
+	if err != nil {
+		return cciptypes.HomeChainConfig{}, err
```
wrap the errors using `fmt.Errorf("some msg: %w", err)`

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+	return r.homeChainConfig.NodeSupportedChains[oracleID].Supported
+}
+
+func (r *HomeChainConfigPoller) fetchLatestConfig(ctx context.Context) (cciptypes.HomeChainConfig, error) {
```
This function has two responsibilities.
Get config and set config in the state of the poller.
Either rename it or make it have one.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func (r *HomeChainConfigPoller) Close(ctx context.Context) error {
+	return nil
```
I think we should stop the work that `startPolling` is doing when `Close` is called.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+func (r *HomeChainConfigPoller) convertOnChainConfigToHomeChainConfig(capabilityConfigs []cciptypes.OnChainCapabilityConfig) (cciptypes.HomeChainConfig, error) {
+	var fChain = make(map[cciptypes.ChainSelector]int)
```
Initialize with
```
fChain := make(...

// instead of

var fChain = make(...
```

asoliman92 committed:
Make the poller thread safe and use mocks properly

asoliman92 committed:
Merge branch 'ccip-develop' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Fix merged imports

asoliman92 committed:
Make HomeChainConfig and Poller work with p2p ids directly

Plugin takes in a map `oracleIdToP2pId` that's used to get the correct p2p id to query and work with the homechain config

asoliman92 committed:
HomeChainPoller interface implements ServiceCtx

This facilitates future migration to be used as a service directly from the jobspec

asoliman92 committed:
Fix naming

asoliman92 committed:
Merge remote-tracking branch 'origin' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Fix import issues and fix tests

asoliman92 commented on core/services/ocr3/plugins/ccip/commit/factory.go:
```diff
@@ -44,15 +46,19 @@ func NewPluginFactory() *PluginFactory {
 
 func (p PluginFactory) NewReportingPlugin(config ocr3types.ReportingPluginConfig,
 ) (ocr3types.ReportingPlugin[[]byte], ocr3types.ReportingPluginInfo, error) {
+	var oracleIDToP2pID map[commontypes.OracleID]cciptypes.P2PID // TODO: Get this from ocr config, it's the mapping of the oracleId index in the DON https://docs.google.com/document/d/1nyGUeterqc9P4ztGmoy2EZCWQzxS5scJ_fQQJyCId7s/edit?disco=AAABLcXrUnA
```
Make sure to use injected ocr3Config into the plugin factory and use it to construct the `oracleIDToP2pID`

asoliman92 committed:
Remove local chainlink-common types

asoliman92 commented on core/services/ocr3/plugins/ccip/internal/reader/ccip.go:
```diff
@@ -8,6 +8,7 @@ import (
 	"time"
 
 	"github.com/smartcontractkit/chainlink-common/pkg/types"
+
```
```suggestion

```

asoliman92 committed:
Update core/services/ocr3/plugins/ccip/internal/reader/ccip.go

asoliman92 committed:
Nits

asoliman92 committed:
Remove unused mock

asoliman92 committed:
Merge branch 'refs/heads/ccip-develop' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Merge remote-tracking branch 'origin' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Fix tests after merging

I suspect that the e2e_test emptyOutcome test is not setup properly on master. Need to double-check.

asoliman92 committed:
Fix tests

asoliman92 committed:
Fix home_chain_test

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -23,53 +23,45 @@ import (
-			}
-		}
-		knownSourceChains = knownSourceChains.Union(mapset.NewSet(sources...))
+	// Call to Start doesn't fail if it's already started
+	err := homeChainPoller.Start(ctx)
+	if err != nil {
+		return nil
 	}
```
Don't start the poller here, it will be started outside of the plugin. Can inject it as a dependency though

asoliman92 replied:
It's already injected as a dependency. This is like a safeguard if it wasn't called earlier to be called here as it won't actually start a new one if it was already started. Do you think it's better to check for the Health and if it wasn't healthy to stop running the plugin completely?

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (r *HomeChainConfigPoller) Start(ctx context.Context) error {
+
```
Unnecessary newline

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+func (r *HomeChainConfigPoller) Start(ctx context.Context) error {
+
+	r.mutex.Lock()
+	if r.backgroundCancel != nil {
+		r.lggr.Errorw("Polling already started")
+		// We don't want to return an actual error here as it's already working as expected
+		return nil
+	}
```
You can use `StartStopOnce` to handle all possible start <-> stop state transitions

makramkd replied:
Means you can get rid of your mutex

asoliman92 replied:
I checked how the `StateMachine` (the newer name for `StartStopOnce` is being used. We still need to have a stop channel and then stop it in `Close` for it to work. I can do it for consistency across our repos but it's not less code or clearer than this.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+	}
+
+	r.lggr.Infow("Start Polling HomeChainConfig")
+	go func() {
+		ticker := time.NewTicker(12 * time.Second)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				r.lggr.Infow("Polling stopped")
+				return
+			case <-bgCtx.Done():
+				r.lggr.Infow("Polling stopped")
+				return
+			case <-ticker.C:
+				_ = r.fetchAndSetConfig(ctx)
+			}
+		}
+	}()
```
Put this `func` in a method

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				r.lggr.Infow("Polling stopped")
+				return
```
The `ctx` given to `Start` should not be used like this, there's no guarantee it'll be long lived

dimkouv replied:
if that's the case probably need to do:
```
bgCtx, cancelFunc := context.WithCancel(context.background())
```

asoliman92 replied:
What's the need of the given context in this case then? The way I see it, that the Start function will be called at launch which means that the context should outlive any other plugin contexts. WDYT?

makramkd reviewed (changes requested):
Only read the PR briefly, will be back to look at it some more

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+	lggr             logger.Logger
+	mutex            sync.RWMutex
+	backgroundCancel context.CancelFunc
+}
```
We want to store the full chain config of all chains in this struct, seems like it only has the FChain and the NodeSupportedChains from homeChainConfig

asoliman92 replied:
It's a byte array, I left it out for now until we understand what we need to map it into exactly.

asoliman92 replied:
For the OCR3 config, It's better to add in another PR IMO. This PR is already full of changes. Any objections to that?

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+func (r *HomeChainConfigPoller) Close() error {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.backgroundCancel == nil {
+		return errors.New("Cancel function not set")
+
+	}
+	r.backgroundCancel()
+	return nil
```
If you use `StartStopOnce` you can call `StopOnce` here instead

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -126,11 +125,16 @@ func (p *Plugin) Observation(ctx context.Context, outctx ocr3types.OutcomeContex
 	if err != nil {
 		return types.Observation{}, fmt.Errorf("observe gas prices: %w", err)
 	}
+	consensusObservation := cciptypes.ConsensusObservation{
```
why `consensusObservation` isn't it just some observed config? There is not consensus logic applied to this data.

dimkouv replied:
the observed items now are:
```
(msgs, gasprices, tokenprices, seqnums, consensusObservation)
```

which seems weird

asoliman92 replied:
I wanted to group them so that it's easier to access them and group them in later stages. I'll check how we use it later again, maybe just flattening it is really better.

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -293,7 +306,8 @@ func (p *Plugin) ShouldAcceptAttestedReport(ctx context.Context, u uint64, r ocr
 func (p *Plugin) ShouldTransmitAcceptedReport(ctx context.Context, u uint64, r ocr3types.ReportWithInfo[[]byte]) (bool, error) {
-	if !p.cfg.ObserverInfo[p.nodeID].Writer {
+	homeChainConfig := p.homeChainPoller.GetConfig()
+	if !homeChainConfig.IsSupported(p.oracleIDToP2pID[p.nodeID], p.cfg.DestChain) {
```
should become: `p.supportsChain(p.cfg.destChain)`

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -21,6 +20,9 @@ import (
 // If we cannot observe the dest we return an empty slice and no error..
+// Plugin should be able to observe the max sequence numbers even if it can't read the source chain.
+// This will be used to reach consensus on the observed seqNums and if plugin can read dest chain it should be able to contribute to the observation.
+// If a sequence number is pending (is not on-chain yet), it will be included in the results.
```
That's not true.

asoliman92 replied:
Yes. I wrote it early on when I was still a bit confused about readers/writers..etc

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -137,6 +139,8 @@ func observeTokenPrices(
 }
 
+// Plugin should be able to observe the max sequence numbers even if it can't read the source chain.
+// This will be used to reach consensus on the observed seqNums and if plugin can read dest chain it should be able to contribute to the observation.
```
This comment should not be here. Probably some copy/pasting error.

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -176,6 +180,7 @@ func newMsgsConsensus(
 	// Gather all messages from all observations.
 	msgsFromObservations := make([]cciptypes.CCIPMsgBaseDetails, 0)
 	for _, obs := range observations {
+		//TODO: Make sure that the observer can read the chain
```
This is not required, we checked that in the validation phase

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -446,16 +451,16 @@ func gasPricesConsensus(lggr logger.Logger, observations []cciptypes.CommitPlugi
 // pluginConfigConsensus comes to consensus on the plugin config based on the observations.
 // We cannot trust the state of a single follower, so we need to come to consensus on the config.
 func pluginConfigConsensus(
```
The returned value has `NodeSupportedChains` field empty.

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -566,18 +544,19 @@ func validateObservedSequenceNumbers(msgs []cciptypes.CCIPMsgBaseDetails, maxSeq
-	observerReadChains := mapset.NewSet(observerInfo.Reads...)
 
-	if len(seqNums) > 0 && !observerInfo.Writer {
+	if len(seqNums) > 0 && !supportedChains.IsSupported(destChain) {
```
maybe rename `IsSupported` to `Contains`

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+
+	r.lggr.Infow("Start Polling HomeChainConfig")
+	go func() {
+		ticker := time.NewTicker(12 * time.Second)
```
let's not use hardcoded values

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+) *HomeChainConfigPoller {
+	return &HomeChainConfigPoller{
+		homeChainReader: homeChainReader,
+		lggr:            lggr,
```
initialize all fields and make mutex a pointer

asoliman92 replied:
What's the reason for using a pointer not a mutex directly?

0xnogo replied:
The mutex would be passed by value in different places and can potentially break the synchronization logic.

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+	r.mutex.Lock()
+	if r.backgroundCancel != nil {
+		r.lggr.Errorw("Polling already started")
+		// We don't want to return an actual error here as it's already working as expected
```
You can return an error, client should handle it

dimkouv replied:
You never unlock, this will lead to big issues

asoliman92 replied:
good catch!

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
```
no need to check `ctx`, `bgCtx` includes it

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+				r.lggr.Infow("Polling stopped")
+				return
+			case <-ticker.C:
+				_ = r.fetchAndSetConfig(ctx)
```
we'll miss all this errors?

asoliman92 replied:
I'm logging inside the `fetchAndSetConfig`, would logging here be better in this case?

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+	return r.homeChainConfig
+}
+
+func (r *HomeChainConfigPoller) GetSupportedChains(p2pID cciptypes.P2PID) mapset.Set[cciptypes.ChainSelector] {
```
this method is not called by anyone

asoliman92 replied:
Called here: https://github.com/smartcontractkit/ccip/blob/8afb0844b3ac49f34f703ed73d4b9f5f4ca0017f/core/services/ocr3/plugins/ccip/commit/plugin.go#L103-L103

dimkouv commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,152 @@
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	if r.backgroundCancel == nil {
+		return errors.New("Cancel function not set")
```
Error string should not be capitalized or end with punctuation mark

dimkouv replied:
non necessary newline

asoliman92 committed:
WIP simplify chain configs

asoliman92 committed:
Fixing tests

asoliman92 commented on core/services/ocr3/plugins/ccip/commit/plugin_e2e_test.go:
```diff
@@ -33,12 +37,13 @@ func TestPlugin(t *testing.T) {
-			nodes:       setupEmptyOutcome(ctx, t, lggr),
-			expErr:      func(t *testing.T, err error) { assert.Equal(t, testhelpers.ErrEmptyOutcome, err) },
-		},
+		//TODO: I think this test is not correct, it should error because no readers are set up for it.
```
@dimkouv ?

asoliman92 committed:
Cleaning up

asoliman92 committed:
WIP move bgCtx outside for HomeChainPollerConfig

asoliman92 committed:
Using StateMachine for running once

asoliman92 committed:
Fix empty outcome test

asoliman92 committed:
Merge branch 'ccip-develop' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
fix linting

asoliman92 committed:
Remove log when stopping (fixing linting CI)

makramkd commented on core/services/ocr3/plugins/ccip/commit/factory.go:
```diff
@@ -48,22 +51,24 @@ func NewPluginFactory() *PluginFactory {
 		config.OracleID,
+		oracleIDToP2pID,
 		cciptypes.CommitPluginConfig{},
 		nil,
 		onChainTokenPricesReader,
 		nil,
 		nil,
 		nil,
+		nil,
```
Can we add comments to these `nil` fields so we know what they are (until they get fully specified):

```suggestion
		nil, /* blah item 1 */
		onChainTokenPricesReader,
		nil, /* blah item 2 */
		nil,
		nil,
		nil,
```

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -325,10 +325,47 @@ func (p *Plugin) Close() error {
+	supportedChains, err := p.homeChainPoller.GetSupportedChainsForPeer(p2pID)
+	if err != nil {
+		p.lggr.Warnf("error getting supported chains: %w", err)
+		// Return empty set
+		// TODO: Should we completely fail?
+		return mapset.NewSet[cciptypes.ChainSelector](), nil
```
If the error is not nil don't return nil for it, return `nil, fmt.Errorf("...: %w", err)`

asoliman92 replied:
Do we want to fail though in case no supportedChains or just return empty set? returning empty set will make the flow smoother during the plugin run.

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin_functions.go:
```diff
@@ -472,51 +468,8 @@ func pluginConfigConsensus(
 	}
-	consensusCfg.FChain = consensusFChain
-
-	// Come to consensus on what the feeTokens are.
```
@dimkouv is it ok that we removed this?

makramkd replied:
Will this be added back at some point in the future?

asoliman92 replied:
I don't see a reason for adding the `consensusObserverInfo` back. For `consensusFeeTokens ` it might be needed in the future but we're not using it at the moment. It should be easy to do it once it's needed. Adding code that's not used now (at least from unit tests perspective), makes it harder to follow. When I asked @dimkouv about this he mentioned that he just added all the things that he thinks we might need in the future and if they're not used we can remove.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+	mutex  *sync.RWMutex
+
+	stopCh services.StopChan
+	services.StateMachine
```
Style nit: put embedded objects at the beginning of the struct decl

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+}
+
+func (r *HomeChainConfigPoller) Start(ctx context.Context) error {
+	err := r.fetchAndSetConfigs(ctx)
+	if err != nil {
+		r.lggr.Errorw("Initial fetch of on-chain configs failed", "err", err)
+		return fmt.Errorf("initial fetch of on-chain configs failed")
+	}
```
I don't think we should fail to start if we couldn't fetch it the first time. We should just go into the `poll` loop and keep retrying on failure

asoliman92 replied:
good point

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			err := r.fetchAndSetConfigs(ctx)
+			if err != nil {
```
Style nit: saves one line

```suggestion
			if err := r.fetchAndSetConfigs(ctx); err != nil {
```

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+		return err
+	}
+	if len(chainConfigInfos) == 0 {
+		r.lggr.Errorw("no on chain configs found")
+		return fmt.Errorf("no on chain configs found")
```
This is not necessarily an error, there can legitimately be no chain configs onchain yet (i.e for a fresh deployment)

makramkd replied:
You can log a warning though

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+		return err
+	}
+	r.lggr.Infow("Setting ChainConfig")
+	r.mutex.Lock()
+	r.chainConfigs = homeChainConfigs
+	r.nodeSupportedChains = createNodesSupportedChains(homeChainConfigs)
+	r.knownSourceChains = createKnownChains(homeChainConfigs)
+	r.fChain = createFChain(homeChainConfigs)
+	r.mutex.Unlock()
```
Cleaner to put this in a func that does the state updates and do a `r.mutex.Lock()` and `defer r.mutex.Unlock()`, these lines kinda blend together atm.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	if len(r.chainConfigs) == 0 {
+		return nil, errors.New("no chain configs found")
```
This is not an error

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+		knownSourceChains.Add(chain)
+	}
+	if knownSourceChains.Cardinality() == 0 {
+		return nil, fmt.Errorf("no known chain configs")
```
Not an error

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	if len(r.fChain) == 0 {
+		return nil, fmt.Errorf("no FChain values found")
```
Don't think this is an error

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+func (r *HomeChainConfigPoller) Ready() error {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	if len(r.chainConfigs) == 0 {
+		return errors.New("no chain configs found")
+	}
```
I don't think this is an error state

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+
+func (r *HomeChainConfigPoller) convertOnChainConfigToHomeChainConfig(capabilityConfigs []ChainConfigInfo) (map[cciptypes.ChainSelector]cciptypes.ChainConfig, error) {
+	chainConfigs := make(map[cciptypes.ChainSelector]cciptypes.ChainConfig)
+	//iterate over configs
```
Unnecessary comment

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,258 @@
+	return chainConfigs, nil
+}
+
+type OnChainConfig struct {
```
I dislike this name, onchainConfig already means something in the OCR context and it could get confusing

asoliman92 replied:
What are your suggestions? Maybe `HomeChainConfigMapper`?

makramkd reviewed (commented):
Need to think about the health check a bit, lack of chain configs is not necessarily an error state

asoliman92 committed:
Update core/services/ocr3/plugins/ccip/internal/reader/home_chain.go

Co-authored-by: Makram <makramkd@users.noreply.github.com>

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -325,10 +325,47 @@ func (p *Plugin) Close() error {
+	if err != nil {
+		p.lggr.Warnf("error getting supported chains: %w", err)
+		// Return empty set
+		// TODO: Should we completely fail?
```
yes

dimkouv commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -325,10 +325,47 @@ func (p *Plugin) Close() error {
+	return supportedChains, nil
+}
+
+func (p *Plugin) getChainConfig() (cciptypes.ChainConfig, error) {
```
i believe this method is not needed you can call poller.getchainconfig() it's the same thing

asoliman92 replied:
It's just a shortcut, calling it from a plugin means get the chain config for the destination chain that this plugin is working with. I might change the name to `getDestChainConfig` to be clearer.

asoliman92 committed:
Addressing PR comments

asoliman92 committed:
Function rename

asoliman92 committed:
Moving HomeChainPoller from CL-common here until we figure out how it will fit in the bigger picture and whether it's needed in CL-common or not.

Relying on cl-common version that doesn't have MsgHasher changes for multiramps

asoliman92 committed:
Merge branch 'ccip-develop' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Moving ChainConfig from cl-common to here

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -7,9 +7,13 @@ import (
 	"time"
 
 	mapset "github.com/deckarep/golang-set/v2"
+
+	"github.com/smartcontractkit/ccipocr3/internal/reader"
+
 	"github.com/smartcontractkit/libocr/commontypes"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
 	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	libocrtypes "github.com/smartcontractkit/libocr/ragep2p/types"
```
Can we group the imports correctly? All ccipocr3 imports should be next to each other at the bottom of the import statement since its a local import.

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -102,8 +92,13 @@ func (p *Plugin) Query(_ context.Context, _ ocr3types.OutcomeContext) (types.Que
 //	We discover the token prices only for the tokens that are used to pay for ccip fees.
 //	The fee tokens are configured in the plugin config.
 func (p *Plugin) Observation(ctx context.Context, outctx ocr3types.OutcomeContext, _ types.Query) (types.Observation, error) {
+	destSupportedChains, err := p.supportedChains()
```
This is just supported chains right? Why is the var called `destSupportedChains`?

asoliman92 replied:
that's a typo, it's supportedChains not destSupportedChains. Thanks for noticing.

makramkd commented on core/services/ocr3/plugins/ccip/commit/plugin_e2e_test.go:
```diff
@@ -5,19 +5,24 @@ import (
 
+	"github.com/smartcontractkit/ccipocr3/internal/reader"
+
+	cciptypes "github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+	libocrtypes "github.com/smartcontractkit/libocr/ragep2p/types"
+	"github.com/stretchr/testify/mock"
 	"github.com/stretchr/testify/require"
 
 	"github.com/stretchr/testify/assert"
 
 	"github.com/smartcontractkit/libocr/commontypes"
-	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
-	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
 
 	"github.com/smartcontractkit/ccipocr3/internal/libs/testhelpers"
 	"github.com/smartcontractkit/ccipocr3/internal/mocks"
 
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+
```
Ditto here re: imports.

makramkd replied:
This should be failing the golangci-lint check

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+	fChain map[cciptypes.ChainSelector]int
+	lggr   logger.Logger
+	mutex  *sync.RWMutex
+	// How frequent will the poller fetch the chain configs
```
```suggestion
	// How frequently the poller fetches the chain configs
```

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+}
+
+func (r *HomeChainConfigPoller) Start(ctx context.Context) error {
+	err := r.fetchAndSetConfigs(ctx)
+	if err != nil {
+		// Just log, don't return error as we want to keep polling
+		r.lggr.Errorw("Initial fetch of on-chain configs failed", "err", err)
+	}
```
What's the point of doing this here if you're going to do it in `poll` anyway?

asoliman92 replied:
To not wait for the ticker and start with a correct state if possible. It also makes testing easier.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+func (r *HomeChainConfigPoller) poll() {
+	ctx, cancel := r.stopCh.NewCtx()
+	defer cancel()
+	ticker := time.NewTicker(r.pollingInterval * time.Second)
```
`pollingInterval` is already a `time.Duration`. Multiplying it again by a `time.Second` will not get you the results you seek; `time.Duration` is denominated in nanoseconds, so `time.Second` is actually `1e9` in raw integer terms, so if you specified `pollingInteval` as the logical 12 seconds (block time on ethereum mainnet) this ends up being `1e9 * 12e9 == 12e18` which is actually larger than the largest `int64`.

This is unless you are setting `pollingInterval` to be something like `1`, essentially a multiplier of the `time.Second` value, but this is not the correct usage of the `time.Duration` API. It'll just lead to confusion and bugs.

So this should just be 
```suggestion
	ticker := time.NewTicker(r.pollingInterval)
```

And the `pollingInterval` should be a meaningful `time.Duration` value (e.g `12 * time.Second`, maybe better as a constant instead of a struct field tbh)

asoliman92 replied:
Good catch about the Duration part. I started initially with `12 * time.Second`. @dimkouv had a comment to make it a variable. You don't see this as a changing parameter or something to experiment with in the future?

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+	}
+	if len(chainConfigInfos) == 0 {
+		// That's a legitimate case if there are no chain configs on chain yet
+		r.lggr.Warnw("no on chain configs found")
```
Maybe return early here to avoid the misleading `Setting ChainConfig` log below

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+	services.StateMachine
+
+	homeChainReader types.ContractReader
+	// gets updated by the polling loop
+	chainConfigs map[cciptypes.ChainSelector]ChainConfig
+	// mapping between each node's peerID and the chains it supports. derived from chainConfigs
+	nodeSupportedChains map[libocrtypes.PeerID]mapset.Set[cciptypes.ChainSelector]
+	// set of chains that are known to CCIP, derived from chainConfigs
+	knownSourceChains mapset.Set[cciptypes.ChainSelector]
+	// map of chain to FChain value, derived from chainConfigs
+	fChain map[cciptypes.ChainSelector]int
```
Suggestion: it feels to me that it'd be clearer what the state being tracked by this object is if we collect them all into one object and have that in this struct. For example:

```golang
type state struct {
    // gets updated by the polling loop
	chainConfigs map[cciptypes.ChainSelector]ChainConfig
	// mapping between each node's peerID and the chains it supports. derived from chainConfigs
	nodeSupportedChains map[libocrtypes.PeerID]mapset.Set[cciptypes.ChainSelector]
	// set of chains that are known to CCIP, derived from chainConfigs
	knownSourceChains mapset.Set[cciptypes.ChainSelector]
	// map of chain to FChain value, derived from chainConfigs
	fChain map[cciptypes.ChainSelector]int
}

type HomeChainConfigPoller struct {
    services.StateMachine

    stopCh services.StopChan

	homeChainReader types.ContractReader
    state state
// ...
}
```

This makes it clearer that state updates need to be atomic (i.e on each poll the whole state will be updated, not just part of it).

This might mean you're even able to move the `mutex` to this state object (though not required).

asoliman92 replied:
Agreed.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	return r.chainConfigs, nil
+
```
Unnecessary newline

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,271 @@
+func (r *HomeChainConfigPoller) HealthReport() map[string]error {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	if len(r.chainConfigs) == 0 {
+		return map[string]error{"ChainConfig": errors.New("no chain configs found")}
+	}
```
Don't think this is right. Here's a suggestion for health report: if the last 10 polls failed, then we are unhealthy.

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,178 @@
+		configPoller := NewHomeChainConfigPoller(
+			nil,
+			logger.Test(t),
+			1,
```
This is not the right usage of the time.Duration API, `1` here is in nanoseconds. You should pass something meaningful to time.Duration in the form usually of existing constants and a multiplier. e.g if you want 12 seconds, you pass in `12 * time.Second`.

asoliman92 replied:
My bad. I got them mixed.

asoliman92 committed:
Update core/services/ocr3/plugins/ccip/internal/reader/home_chain.go

Co-authored-by: Makram <makramkd@users.noreply.github.com>

asoliman92 committed:
Make an internal state for home chain poller

* Use duration parameter correctly
* Fail health check when failing to poll 10 times in a row

asoliman92 committed:
Merge branch 'ccip-develop' into feat/CCIP-2466-homechain-config-sync

asoliman92 committed:
Fix linting errros

makramkd commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+func (r *HomeChainConfigPoller) setState(chainConfigs map[cciptypes.ChainSelector]ChainConfig) {
+	r.mutex.Lock()
+	defer r.mutex.Unlock()
+	s := &r.state
+	s.chainConfigs = chainConfigs
+	s.nodeSupportedChains = createNodesSupportedChains(chainConfigs)
+	s.knownSourceChains = createKnownChains(chainConfigs)
+	s.fChain = createFChain(chainConfigs)
```
You can just update `r.state` directly like `r.state.chainConfigs = ...` etc. `r` is already a pointer receiver so any member updates happen on object memory

winder commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -23,53 +27,39 @@ import (
-	readableChains mapset.Set[cciptypes.ChainSelector]
-	// knownSourceChains is the set of chains that the plugin knows about.
-	knownSourceChains mapset.Set[cciptypes.ChainSelector]
+	homeChainPoller reader.HomeChainPoller
```
nit: consider renaming this to `reader.HomeChain`. The fact that it's polling for changes is not really important from the caller perspective. In fact, for testing we would likely substitute a static implementation of the `reader.HomeChain` interface.

asoliman92 committed:
Merge branch 'ccip-develop' into feat/CCIP-2466-homechain-config-sync

winder commented on core/services/ocr3/plugins/ccip/commit/plugin.go:
```diff
@@ -188,10 +196,6 @@ func (p *Plugin) ValidateObservation(_ ocr3types.OutcomeContext, _ types.Query,
 		return fmt.Errorf("validate gas prices: %w", err)
 	}
 
-	if err := obs.PluginConfig.Validate(); err != nil {
-		return fmt.Errorf("validate plugin config: %w", err)
-	}
```
This should still be validated, it looks like some of the values are not part of the home chain. (i.e. ScanBatchSize)

winder replied:
Although, I wonder why it's done during observation rather than initialization...

asoliman92 replied:
Observations don't have PluginConfig anymore

winder commented on core/services/ocr3/plugins/ccip/commit/plugin_e2e_test.go:
```diff
@@ -181,48 +185,104 @@ func TestPlugin(t *testing.T) {
-		newNode(ctx, t, lggr, 1, cfg),
-		newNode(ctx, t, lggr, 2, cfg),
-		newNode(ctx, t, lggr, 3, cfg),
+	chainConfigInfos := []reader.ChainConfigInfo{
```
A general comment on these tests: there is a lot of mocking and magic numbers which makes it hard to see whats actually being tested...

asoliman92 replied:
I agree with that. I'll address these in another PR 👍

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	fChain map[cciptypes.ChainSelector]int
+}
+
+type HomeChainConfigPoller struct {
```
Can this be private?

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	homeChainReader types.ContractReader,
+	lggr logger.Logger,
+	pollingInterval time.Duration,
+) *HomeChainConfigPoller {
```
Return the interface type

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	s := r.state
+	if _, ok := s.chainConfigs[chainSelector]; !ok {
+		return ChainConfig{}, fmt.Errorf("chain config not found for chain %v", chainSelector)
+	}
+	return s.chainConfigs[chainSelector], nil
```
nit: you can do this with a single lookup
```suggestion
	if chainConfig, ok := s.chainConfigs[chainSelector]; ok {
		return chainConfig, nil
	}
	return ChainConfig{}, fmt.Errorf("chain config not found for chain %v", chainSelector)
```

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	return chainConfigs, nil
+}
+
+// HomeChainConfigMapper This is a 1-1 mapping between the config that we get from the contract to make se/deserializing easier
+type HomeChainConfigMapper struct {
+	Readers []libocrtypes.PeerID `json:"readers"`
+	FChain  uint8                `json:"fChain"`
+	Config  []byte               `json:"config"`
+}
+
+// ChainConfigInfo This is a 1-1 mapping between the config that we get from the contract to make se/deserializing easier
+type ChainConfigInfo struct {
+	// Calling function https://github.com/smartcontractkit/ccip/blob/330c5e98f624cfb10108c92fe1e00ced6d345a99/contracts/src/v0.8/ccip/capability/CCIPCapabilityConfiguration.sol#L140
+	ChainSelector cciptypes.ChainSelector `json:"chainSelector"`
+	ChainConfig   HomeChainConfigMapper   `json:"chainConfig"`
+}
+
+// ChainConfig will live on the home chain and will be used to update chain configuration like F value and supported nodes dynamically.
+type ChainConfig struct {
+	// FChain defines the FChain value for the chain. FChain is used while forming consensus based on the observations.
+	FChain int `json:"fChain"`
+	// SupportedNodes is a map of PeerIDs to SupportedChains.
+	SupportedNodes mapset.Set[libocrtypes.PeerID] `json:"supportedNodes"`
+	// Config is the chain specific configuration.
+	Config []byte `json:"config"`
+}
```
Do these need to be public?

asoliman92 replied:
I use them in the e2e tests to simulate the polling fully in the commit plugin

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+	assert.Errorf(t, errors["HomeChainConfigPoller"], "polling failed %d times in a row", MaxFailedPolls)
+}
+func Test_ConvertOnChainConfigToHomeChainConfig(t *testing.T) {
+	var tests = []struct {
```
nit: no need for table driven tests if there's just one test

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+			},
+		},
+	}
+	for _, tc := range tests {
```
```suggestion
	for _, tc := range tests {
		tc := tc
```

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+	_ = configPoller.Close()
+	assert.Equal(t, 1, len(errors))
+	assert.Errorf(t, errors["HomeChainConfigPoller"], "polling failed %d times in a row", MaxFailedPolls)
+}
```
```suggestion
}

```

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+		logger.Test(t),
+		50*time.Millisecond,
+	)
+	_ = configPoller.Start(context.Background())
```
Even if it cannot return an error now, it should check. (also, this function can return an error)
```suggestion
	err := configPoller.Start(context.Background())
	require.NoError(t, err)
```

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	return chainConfigInfo, err
+}
+
+func (r *HomeChainConfigPoller) convertOnChainConfigToHomeChainConfig(capabilityConfigs []ChainConfigInfo) (map[cciptypes.ChainSelector]ChainConfig, error) {
```
This appears to be a pure function
```suggestion
func convertOnChainConfigToHomeChainConfig(capabilityConfigs []ChainConfigInfo) (map[cciptypes.ChainSelector]ChainConfig, error) {
```

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain.go:
```diff
@@ -0,0 +1,285 @@
+	return nodeSupportedChains
+}
+
+func (r *HomeChainConfigPoller) fetchOnChainConfig(ctx context.Context) ([]ChainConfigInfo, error) {
```
This function doesn't reduce the number of error checks required by the caller, I'd probably just keep it in fetchAndSetConfigs rather than making it a separate function

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+		},
+	}
+	for _, tc := range tests {
+		configPoller := NewHomeChainConfigPoller(
+			nil,
+			logger.Test(t),
+			1*time.Second,
+		)
```
You can skip this setup by making convertOnChainConfigToHomeChainConfig a pure function

winder commented on core/services/ocr3/plugins/ccip/internal/reader/home_chain_test.go:
```diff
@@ -0,0 +1,210 @@
+	ctx := context.Background()
+	_ = configPoller.Start(ctx)
+	_ = configPoller.Close()
+	time.Sleep(100 * time.Millisecond)
```
This is smaller than the polling interval. This is testing that values are fetched during `Start`, not because of the polling.

winder reviewed (approved):
Mostly nits and style suggestions, I don't think any of the feedback is strictly blocking. So approved pending your consideration of feedback.
